storage/heap/CMakeLists.txt

```diff
@@ -27,4 +27,8 @@ IF(WITH_UNIT_TESTS)
 
   ADD_EXECUTABLE(hp_test2 hp_test2.c)
   TARGET_LINK_LIBRARIES(hp_test2 heap mysys dbug strings)
-ENDIF()
\ No newline at end of file
+  IF(CMAKE_SYSTEM_NAME MATCHES AIX)
+    TARGET_LINK_OPTIONS(hp_test1  PRIVATE "-Wl,-berok")
```

Comment:
This looks a bit repetitive. Can [LINK_OPTIONS](https://cmake.org/cmake/help/v3.15/prop_tgt/LINK_OPTIONS.html) be set globally that achieves the same result?

Reply:
With `-Wl,-berok`, linker will not complain if a symbol is missing. It can be useful for unit tests (because some symbols are loaded during execution with dlopen). However, it is dangerous in the generic case because an error during compilation will be delayed to execution.
I have added `-Wl,-berok` to many unit tests by modifying the `MY_ADD_TESTS` macro. However, some tests does not use it, e.g. **hp_test1/2** in `storage/heap/CMakeLists.txt`.
I think that add `-Wl,-berok` to LINK_OPTION will decrease quality and can cause an unexpected crash, especially when a plugin is load, in a future version.

Reply:
Windows does not have this option, to allow unresolved symbols, and it links fine. Linux et all do not allow that in executables. In the past, it was not needed (granted, my last experience with AIX is a decade ago). So what makes AIX here so special? What is being dlopened?

Reply:
> In the past, it was not needed on AIX either
Because MariaDB used `-bexpall`, that has side effect. See https://github.com/MariaDB/server/pull/1515/files#r417221418 .

Reply:
And CMake added `-G` (so `-Wl,-brtl` and `-Wl,-berok`). These flags has other side effect, including performance issue and trouble when compiling C++. But is easier to compile with them.

Reply:
I'm actually in favour of bexpall/bexpfull, compared to all this it seems like much less trouble. It worked, I personally tested it (was a decade ago, but I hope IBM stays compatible to itself).
I can, sort of, understand the trouble you describe when an executable would export symbols from system *static* libraries.
But I doubt you can re-export symbols from system "shared" libraries, so you could use them maybe to avoid the effect.

What are you trying to do (export all symbols from chosen static libraries), is the same thing that I have done on Windows, back in dark CMake times, 10 years ago, when there was no way to export al symbols from executable, so I can relate, but I would not repeat anything that a ) CMake and b) OS linker both can do.

Reply:
About `bexpall`, it can export symbol of other libraries, including system. As MariaDB provides libraries, it can pollute other (they will search system symbol in MariaDB and not the system itself). This behavior is really bad.

> But I doubt you can re-export symbols from system "shared" libraries, so you could use them maybe to avoid the effect.

See Bullfreeware news (http://www.bullfreeware.com/newsPage), Compatibility OpenSSL Removal. "Our OpenSSL was polluted by -bexpall.". In fact, old openssl of Bullfreeware export symbols of std lib, eg `strcmp`. Do not doubt, download eg this one,  http://www.bullfreeware.com/pkg?id=3952 , extract libcrypto.a and 
````
dump -Tv -X32 libcrypto.a | grep strcmp
[416]   0x20014de0    .data      EXP     DS SECdef        [noIMid] strcmp
````

> What are you trying to do (export all symbols from chosen static libraries), is the same thing that I have done on Windows, back in dark CMake times, 10 years ago

CMake support of AIX is partial in my point of vue.

Reply:
I do not see strcmp as example of being too bad.  The only thing you need to care of is libmariadb (connector/c), because it can be used outside of the server. and Connector/C does have its list of exported symbols that you can use. that list is btw the same for the embedded library.

Reply:
I'm also rather curious, how does linker export symbols from system libraries (other than system libraries would be linked statically)

Reply:
Overall, I think following strategy should work
1. you export *all symbols* from mariadbd executable, to use for plugins. Take -bexpall, bexpfull , whatever appropriate. I know it is not nice, but you're emulating -rdynamic on Linux 
, which is known to work. You can use that for for mariadbd only
 
2. You use -Wl,-berok as link option for plugins. I know it is not nice,  but you're emulating Linux behaviour, which is known to work

3. You adopt MYSQL_PLUGIN_EXPORT (lookup in the source) to work with AIX compiler.
It is , quite probably, some visibility(default) . Anyway, this should emulate __declspec(dllexport) on Windows.

4. Take a look how .def file is generated, for Connector/C . The symbols are already known, so you can use that for the AIX export file, too.

With all this in place, I think you'd not need any code to extract symbols from static libraries, and I think it would be much cleaner this way.

Reply:
And, if unit tests do not link , you'd need to add appropriate static library. I see no point in
-Wl,-berok when linking *executables*. it is appropriate for what CMake calls MODULE, but for executables, all symbols must be resolved, without resorting to hacks.

Reply:
Hi Vaintroub, 

I'm another guy working on AIX and trying to improve GCC and the whole OpenSource on it. I'm just here to say a few things to support Etienne point of view. 

> you export all symbols from mariadbd executable, to use for plugins. Take -bexpall, bexpfull , whatever appropriate. I know it is not nice, but you're emulating -rdynamic on Linux, which is known to work. You can use that for for mariadbd only

We have completely forbidden these two flags, `-bexpall` and `-bexpfull`,  in our recent RPMs in order to avoid the huge problems we had in the past, even trigger by a small  `strcmp` in OpenSSL. Even if `libmariadb` isn't a library used by many different software, we (BullFreeware and IBM) don't want to risk having programs searching `libc` symbols inside another library. This can be a solution for someone working on a project for himself, but that's not a solution when you're trying to build an Operating System with backward compatibility. 

> You use -Wl,-berok as link option for dynamic plugins, in cmake/plugin.cmake. I know it is not nice, but you're emulating Linux behaviour, which is known to work

Even if it seems a good solution for plugins at first, `-berok` is not supposed to be used without `-brtl` or `-G` (and using them is out of question because of the backward compatibility). Without them, it'll simply silence the linker errors about undefined symbols. If indeed a symbol was missing, launching the test will, most of the time, result into a beautiful SIGILL, because the program tried to call an undefined symbol. It's not that debugging it is complicated. But, as developer, I think you can understand we'd rather have "X symbol is missing" than "Illegal instruction". 

Please, do remember that AIX linker and AIX loader are both capricious if not stubborn. Etienne has found a solution which is creating plugins without having `-brtl` or `-G`, and thus can be backward compatible. It might be improved for sure, but using `-bexpall` mixed with `-berok` is not a improvement.

Reply:
Oh,, maybe you can enlighten me on how comes that symbols from standard library are exported? I can share my past experience, and that was with Windows, and this happened when
I linked C runtime statically. I can't imagine you want this in this project, as I have a hard time imagining any other mechanism (apart maybe something like DLL redirection, which is harmless) by which you'd export standard symbols.  Does it -bexpfull export instrinsics or somesuch? I'm genuinely interested.

Hopefully libmariadb could be a  library used by a lot of software. You can link and compile it however you want, with/without brtls and -G. There is an API , and an official export list you can use, like Windows does it.

mariadbd however is an executable, not really comparable to either libmariadb, or OpenSSL.
Nobody links with it. Its exported symbols are for its own plugins.

And even if you ban these wonderful options by your own linker, you can still look out for better alternatives.
CMake as of today does not export symbols from static libraries, this is true.
It may very well exports them, if  you add $<TARGET_OBJECTS:mysys> to the list of source files, rather than target_link_libraries(mariadbd mysys) . There is a proof for that in current 10.5 source tree, in sql/CMakeLists.txt. Did you try that? Did it work?

What is wrong with -brtl, *for server plugins*.? They are only loaded at runtime, they are only used with dlopen and dlsym. server plugins are a poster child for -brtl, from the  IMO, or not?

Server plugins do not have a specific ABI,   they are only useful together with the server, from the same build.  Though I would not mind, if your server plugins link with the  mariadbd executable, using later as "shared object" (like Macintosh does, or did in the past). In case your linker allows that, that would be the easiest option.

Reply:
> Oh,, maybe you can enlighten me on how comes that symbols from standard library are exported? I can share my past experience, and that was with Windows, and this happened when I linked C runtime statically. I can't imagine you want this in this project, as I have a hard time imagining any other mechanism (apart maybe something like DLL redirection, which is harmless) by which you'd export standard symbols. Does it -bexpfull export instrinsics or somesuch? I'm genuinely interested.

We already this discussion with the Cmake guys: https://gitlab.kitware.com/cmake/cmake/-/issues/19163 (David Edelson being the head developper of GCC on AIX and he's far more competent than me on these topics). But basically, `-bexpall` isn't as accurate as it's supposed to be: C++ symbols aren't all exported, imported symbols sometimes end up being exported too. I don't know why though, this option was added to be closer to Linux behavior but it was, it seems, botched... 
Using it for executables should work, at least in theory, but with all the weird stuff going on with it, I would rather not using it..  

> Hopefully libmariadb could be a library used by a lot of software. You can link and compile it however you want, with/without brtls and -G. There is an API , and an official export list you can use, like Windows does it.

I wasn't aware of that. It might be a better solution than creating our own indeed. 

> What is wrong with -brtl, for server plugins.? They are only loaded at runtime, they are only used with dlopen and dlsym. server plugins are a poster child for -brtl, from the IMO, or not?

The main problem with `-brtl` is that its changing the way libraries are linked. 
Without it, ld is linking executable and libraries to a member (which is a shared object) inside an archive (eg "libmariadb.a(libmariadb.so.3)"). When the .so compatibility is changed, a new .so can be created and suffixed with here 4. The previously made programs will still be able to load with  "libmariadb.a(libmariadb.so.3)" and the new ones will have  "libmariadb.a(libmariadb.so.4)". Thus the backward compatibility is guaranteed. 
Now, with it, ld will try to link directly to a shared object without version suffix (eg "libmariadb.so"). Thus, backward compatibility can't be guarantee: a new .so version will break previous programs. Note that if the shared object isn't available, it will use the archive instead. 
To understand how complex it's you have to remember that we are working for an operating system. Imagine that at least one plugin is using a system library, eg libz. This plugin with `-brtl` will want to link with libz.so. If it's indeed available, the plugin won't be backward compatible. But if only the archive is available, the link will be ok, and we can used `-brtl`. 
Let's push a bit further, we are using RPM to deliver new packages. RPM when building a package is looking for requires. If a library or an executable is linked with libz.so, thus libz.so will be a require. Thus, our zlib package must provide libz.so otherwise we won't be able to installed this package. 
Finally, imagine that an already made package using `-brtl` for a long long time is central to an operating system, let's say Python. These Python's RPM have a require on libz.so, so it's not backward compatible but we are aware of that, and we are being very careful with that. Thus libz.so is present on our systems as long as Python is installed (which is because of yum/dnf). Thus, the mariadb plugin which is built with "-lz" will be linked with libz.so, so will not be backward compatible. 
I hope this is a bit clearer. This is a complex subject and we are trying to avoid the errors made in the past. "-brtl" must be used very very carefully and we will try to remove it totally from our system one day. That's why we don't want it on MariaDB as long as there is another way and there is one. 

TL;DR: `-brtl` is aimed to reproduce Linux behavior which isn't backward compatible. AIX is design to preserve backward compatibility and it's its strenght. Thus, `-brtl` must be avoided, if possible.

Reply:
> It may very well exports them, if you add $<TARGET_OBJECTS:mysys> to the list of source files, rather than target_link_libraries(mariadbd mysys) . There is a proof for that in current 10.5 source tree, in sql/CMakeLists.txt. Did you try that? Did it work?

I have not tried before. A first try do not work. All CMake Windows code is consistent with code in MSVC flag. AIX is consistent with UNIX flag. So, it will require more work.

Moreover, if I correctly understand your idea, you want create TARGET_OBJECTS to have an export file automatically created on AIX. I have checked with an other project, and CMake does not produce more export file if I use TARGET_OBJECTS. I am not sure AIX_EXPORT_ALL_SYMBOLS and WINDOWS_EXPORT_ALL_SYMBOLS have exactly the same effect. As I said, CMake support of AIX is partial only.

Reply:
ok, so again
1. You have an executable. You want to export some symbols from this executable. Your symbols are in static libraries linked to that executable.
What you need to do is 

- ENABLE_EXPORTS on the executable. This is documented to create an export, on AIX specifically, as per
https://cmake.org/cmake/help/v3.16/prop_tgt/ENABLE_EXPORTS.html

- Instead of linking static library, use objects from this library as source files.
i.e do not TARGET_LINK_LIBRARY(mariadbd mysys)
Do 
  ADD_EXECUTABLE(mariadbd ${some_source_files}
    $<TARGET_OBJECTS:mysys>
  ...)
instead.

AIX_EXPORT_ALL_SYMBOLS is in my understanding, the opt-out option, default ON but you might better ask Brad King from CMake, who has access to both AIX, and knows CMake code

So far, there was nothing MSVC specific. WINDOWS_EXPORT_ALL_SYMBOLS is not even about MSVC, clang does this quite the same way,

Reply:
And I think I know why -bexpall does not suffice. If it did not change in the last 10 years, it still would not export symbols with underscore. -bexpfull does

Reply:
So, I reread https://cmake.org/cmake/help/v3.16/prop_tgt/ENABLE_EXPORTS.html more attentively.
What I got from it, is that if you manage to export correct symbols from the static libraries
(hopefully $<TARGET_OBJECTS> technique I described previously, work), then you can link
plugins (which are "loadable modules" in CMake language) to the executable, and this would work

> On AIX, a linker “import file” is created along with the executable to list the exported symbols for import when linking other targets. Loadable modules link to the import file to get the symbols.

Note , that linking with mariadbd is already in place,
https://github.com/MariaDB/server/blob/10.5/cmake/plugin.cmake#L223
 and the only thing you have to do is manage to exporting he functions the correct CMake-ish way.

Reply:
> And I think I know why -bexpall does not suffice. If it did not change in the last 10 years, it still would not export symbols with underscore. -bexpfull does. Therefore we had -bexpfull in the link options, not -bexpall 

Yes that still the case. I actually don't know what a "-bexpfull" on a executable is doing, I'll check that. 

About CMake and plugin, I don't remember exactly if every patches we have made have been merged in CMake. Etienne knows that but is on break this week, so we'll check that next one.

Reply:
Hi vaintroub,
We have some trouble to compile MariaDB 10.5 master. It is related to bugs into AIX gcc (strange toc duplication). I will spend more time to the symbol export issue when the gcc bug will be corrected.

Reply:
Hi,
Sorry for the delay. I have corrected the explicit export using ``dump``. For server, I use the same way than Windows. For libservices, I use an explicit export of symbols by plugins using a similar way than mariadb-connector-c.
As plugins do not import server symbols, ``-berok`` is needed. I do not handle embedded server because it has specific symbols issues, and it is not mandatory.

For your information, ``-bexpall`` is not working, because some symbols (typically those one beginning by a ``_``) are not exported.

Reply:
I know that -bexpall is not working, because we used -bexpfull, which exports also those "some" symbols.
Can you the Windows thing for libservices, too? Otherwise it will be out of data any time when new service is added

this is that we have in service_versions.h

```c++
#ifdef _WIN32
#define SERVICE_VERSION __declspec(dllexport) void *
#else
#define SERVICE_VERSION void *
#endif 
```

what is the  __declspec(dllexport)  for AIX? Is this attribute visibility(default) ?

Reply:
Also, could you reduce the -Wl,berok in the patch? Why is it necessary?

Reply:
For ``bexpfull`` , I have checked, and I obtain a really complex ``ld`` crash. I have not investigated further.

For libservices, I cannot make the same than Windows on AIX. ``__declspec`` does not exist on AIX. The equivalent is to use ``-Wl,-bE`` as I make.

Reply:
For `-Wl,-berok`, I use it on plugins and unittests.
For plugins, it is needed because server symbols are missing without ``-Wl,-bI`` on plugins. And the ``-Wl,-bI`` will require to create an explicit list of symbols. So, I use ``-Wl,-berok`` to compile.
For unittest, I can replace it by a link to the ``sql`` library in most of them (maybe not all, I must check). However, it increases a lot the binary size of all unittest. As we distribute the unittest with mariadb, I want to keep the size reasonable.
All unittests are executed, so we know there is not missing symbols in execution.

Reply:
declspec does not exist on AIX, yes, sure.. On non-MSVC, this is typically handled with visibility attribute, apparently visibility(default), which exist according to  https://www.ibm.com/developerworks/aix/library/au-aix-symbol-visibility-part2/index.html.

Reply:
It is only supported with xlc on AIX (the page you linked is for xlc). With a basic test with gcc, I obtain
````C
#include <stdio.h>

__attribute__ ((visibility ("hidden"))) int times_two(int number)
{
  return number * 2;
}

int main(void)
{
  printf("%d\n", times_two(10));
  return 0;
}
````

````bash
$ gcc main.c
main.c: In function 'times_two':
main.c:7:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
````

Reply:
You should not need berok neither for plugins, nor for unit tests. Tests themselves should not be underlinked, even ELF linker would complain about underlinking for executables. plugins should not be underlinked, because you export all symbols from sql/mysys/etc a la Windows. So, why berok? what if you do not use berok?

What's the goal of the project? You want to compile with gcc? Or do you want to compile with xlc? Would anyone need gcc on AIX, was not it traditionally worse than the native compiler?

Reply:
Tests and plugin are not really underlinked. Symbols are imported from ``mariadbd``. However, AIX ``ld`` requires all imported symbols to be explicitly defined. That's why in the first version of my PR, I have listed (using ``dump``) all symbols in ``mariadbd``, and used the generated file as an import file for plugins (using ``-Wl,-bI`` flag).
Without ``-Wl,-bI``, imported symbols are considered to be missing by plugins, so link fails. That's why I need to use ``-Wl,-berok``.

About ``MY_ADD_TESTS``, it is what I have done. See [this change](https://github.com/MariaDB/server/pull/1515/files#diff-ed5c0552008d49a4d23247426e5a6dcf). Most of the reamining ``-berok`` are related to test using directly ``MY_ADD_TEST`` (without final S) or just ``ADD_EXECUTABLE`` (e.g. myisam tests). I cannot add ``-berok`` to the ``MY_ADD_TEST`` macro, because some tests (e.g. `dbug-t`) are not real executable, so CMake fails to add a flag to this target. Other `-berok` are used by maria or myisam tools, for the same reason that plugins.

The goal of this project is to port MariaDB using GCC on AIX. GCC is the default compiler of the [Toolbox](https://www.ibm.com/support/pages/aix-toolbox-linux-applications-overview) and [Bullfreeware](http://www.bullfreeware.com/) teams. So it is used successfully for lot of OpenSources packages.

Reply:
I would notice if tests would be underlinked, or if they were imported from mariadbd.  There could be one or 2 tests that link to static library "sql" , but most link to specific libraries. mysys is very popular, also dbug and strings. some aria tests link to aria storage engine.

plugins need to link to "server", this is a shared library, both you and Windows are using. The requires symbols are exported, and, if I understood the CMake documentation right, you'll link against an import library. Just like on Windows. 
No unresolved symbols should be there.

Reply:
if you want to test of some tests are a cmake target, you can use IF(TARGET $name)

Reply:
For plugins, link with ``server`` is OK.

I can compile unittests by linking with sql, sql_builtins and perfschema in macro ``MY_ADD_TEST``. However, the unittests size increases a lot (~300 Mo). Adding ``server`` as a dependency for tests or tools is not OK (because of another complex ``ld`` error related to visibility on imported symbols).

Reply:
Which unittests fail at list?

Reply:
Without berok and without specific links, almost all. E.g. storage/perfschema/unittest/pfs-t.
Error is typically ``ld: 0711-317 ERROR: Undefined symbol: handler::engine_name()`` with a lot of missing symbols.

Without berok, and with link to ``server``, almost all again. E.g. unittest/strings/strings-t.
Error is typically
````
ld: 0711-407 ERROR: Symbol .Field_blob::get_ptr(unsigned char const*) const
        Visibility is not allowed on a reference to an imported symbol
````

>  the AIX linker is more picky than elsewhere, and will need symbols that are referenced in object files, but not used in the resulting binary

If I correctly remember, yes.

> this could be fixed in a better way than berok, or linking everything with sql, sql_builtins, all storage engines, and perfschema.

Normally, it is ``-Wl,-bI:import_file.imp``. But for that we need to know all symbols missing. And we must avoid adding symbols from a library linked by the test (I suppose the Visibility error is going from this point).

Reply:
I think strings-t do not need to link to sql, sql_builtins, storage engines etc. it should just link to strings, and maybe mysys.

pfs-t can be slightly special and weird,  and need sql link, because it uses handler class, which is in sql. so you can link it to sql, and sql_builtins, as an exception.

except pfs-t, what are other things that fail without berok?

Reply:
All unittest of maria, myisam and perfschema storage are failing. E.g. ma_control_file-t, ma_pagecache_rwconsist2_1k-t, 
mi_test1, pfs_instr_class-oom-t...

Reply:
I see, so, what was the reason not to add berok to MY_ADD_TESTS() macro,  next to ADD_EXECUTABLE, with proper comments on pecularities of the AIX linker. I'm trying to reduce the diffs in your patch to the minimum, because there will be problems with any of that diff, if anything changes.  On the other hand, if the linker is so fragile, you can add berok to the CMAKE_EXE_LINKER_FLAGS in cmake/os/AIX.cmake. Or will it break anything again?

Reply:
Ok , to summarize the long discussion about -Wl,-berok . I would pretty much prefer  one global flag, LINK_OPTION to the multitude of IF(AIX) scattered through all the CMakeLists.txt. That is, you can't convince to linker to behave. Plugins should not need it, and the fact that executables link on all other platforms, increases the change that they will do on AIX.  and if you run the tests, you'll also that for sure.

Reply:
Another alternative is forcing WITH_UNIT_TESTS to FALSE,  inside cmake/os/AIX.cmake. With a comment, that linker does not behave. Unit tests "-t" , they have rather low importance,  and I think , can be skipped on platforms, where linker does not behave. mtr is the tool of choice for our testing.

Reply:
If we force WITH_UNIT_TESTS to FALSE, maria/myisam tools (e.g. aria_ftdump, aria_dump_log) will continue to need their own  `-berok`. But `-berok` will appears only twice (as it can be removed from cmake/ctest.cmake and it is removed from other files).
If it is not acceptable for you, we can add `-Wl,-berok` to LINK_OPTION.

> you can't convince to linker to behave.

It is one of the best description of AIX ld's, in my opinion.

Reply:
I have removed the `-berok` flags in the `MY_ADD_TEST` macro. It's OK with the "CMAKE_EXE_LINKER_FLAGS" to maria, myisam and perfschema thanks to other modifications.